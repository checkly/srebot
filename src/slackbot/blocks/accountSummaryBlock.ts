import { LIST_ERROR_PATTERNS_ACTION_ID } from "./errorPatternBlock";
import { LIST_FAILING_CHECKS_ACTION_ID } from "./failingChecksBlock";
import { MultipleChecksGoalResponse } from "../../prompts/summarizeCheckGoals";
interface AccountSummaryProps {
  accountName: string;
  passingChecks: number;
  degradedChecks: number;
  failingChecks: number;
  hasIssues: boolean;
  issuesSummary: string;
  failingChecksGoals: MultipleChecksGoalResponse;
  failingCheckIds: string[];
  errorPatterns: { id: string; description: string; count: number }[];
  passingChecksDelta: number;
  degradedChecksDelta: number;
  failingChecksDelta: number;
}

export function createAccountSummaryBlock({
  passingChecks,
  passingChecksDelta,
  degradedChecks,
  degradedChecksDelta,
  failingChecks,
  failingChecksDelta,
  issuesSummary,
  failingChecksGoals,
  failingCheckIds,
  errorPatterns,
  accountName,
}: AccountSummaryProps) {
  const passingChecksDeltaText =
    passingChecksDelta > 0
      ? `(+${passingChecksDelta})`
      : passingChecksDelta < 0
        ? `(${passingChecksDelta})`
        : "";
  const degradedChecksDeltaText =
    degradedChecksDelta > 0
      ? `(+${degradedChecksDelta})`
      : degradedChecksDelta < 0
        ? `(${degradedChecksDelta})`
        : "";
  const failingChecksDeltaText =
    failingChecksDelta > 0
      ? `(+${failingChecksDelta})`
      : failingChecksDelta < 0
        ? `(${failingChecksDelta})`
        : "";

  errorPatterns.sort((a, b) => b.count - a.count);

  return {
    blocks: [
      {
        type: "header",
        text: {
          type: "plain_text",
          text: `Account "${accountName}" last 24h`,
          emoji: true,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `:white_check_mark: *PASSING*: ${passingChecks} ${passingChecksDeltaText}\n :warning: *DEGRADED*: ${degradedChecks} ${degradedChecksDeltaText}\n:x: *FAILING*: ${failingChecks} ${failingChecksDeltaText}`,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `${issuesSummary}`,
        },
      },
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Impact Analysis:*\n${failingChecksGoals.response.length > 0 ? failingChecksGoals.response.map((group, index) => `${index + 1}. *${group.header}*: ${group.description}`).join("\n") : "No failing checks detected in the last 24h."}`,
        },
      },
      ...(errorPatterns.length > 0
        ? [
            {
              type: "section",
              text: {
                type: "mrkdwn",
                text: "*Top 3 Error Patterns:*\n",
              },
            },
            ...errorPatterns.slice(0, 3).flatMap((errorPattern, index) => [
              {
                type: "section",
                text: {
                  type: "mrkdwn",
                  text: `${index + 1}. \`${errorPattern.description}\` (${errorPattern.count} times)`,
                },
              },
            ]),
          ]
        : []),
      {
        type: "divider",
      },
      {
        type: "actions",
        elements: [
          {
            type: "button",
            text: {
              type: "plain_text",
              emoji: true,
              text: "Open Checkly Dashboard",
            },
            url: `https://app.checklyhq.com/`,
          },
          ...(failingCheckIds.length > 0
            ? [
                {
                  type: "button",
                  text: {
                    type: "plain_text",
                    emoji: true,
                    text: "List Failing Checks",
                  },
                  action_id: LIST_FAILING_CHECKS_ACTION_ID,
                  value: failingCheckIds.join(","),
                },
              ]
            : []),
          ...(errorPatterns.length > 0
            ? [
                {
                  type: "button",
                  text: {
                    type: "plain_text",
                    emoji: true,
                    text: "List Error Patterns",
                  },
                  action_id: LIST_ERROR_PATTERNS_ACTION_ID,
                  value: errorPatterns.map((ep) => ep.id).join(","),
                },
              ]
            : []),
        ],
      },
    ],
  };
}
