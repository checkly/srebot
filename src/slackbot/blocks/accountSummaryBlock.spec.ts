import { createAccountSummaryBlock } from "./accountSummaryBlock";

describe("accountSummaryBlock", () => {
  it("renders healthy account state", () => {
    const blocks = createAccountSummaryBlock({
      accountName: "Healthy",
      passingChecks: 50,
      degradedChecks: 0,
      failingChecks: 0,
      passingChecksDelta: 0,
      degradedChecksDelta: 0,
      failingChecksDelta: 0,
      hasIssues: false,
      issuesSummary: "No issues detected in the last 24h.",
      failingChecksGoals: { response: [] },
      failingCheckIds: [],
      errorPatterns: [],
    });

    expect(blocks).toMatchSnapshot();
  });

  it("renders degraded account state", () => {
    const blocks = createAccountSummaryBlock({
      accountName: "Degraded",
      passingChecks: 45,
      degradedChecks: 5,
      failingChecks: 0,
      passingChecksDelta: 0,
      degradedChecksDelta: 0,
      failingChecksDelta: 0,
      hasIssues: true,
      issuesSummary:
        "New degrading or failing checks detected in the last 24h.",
      failingChecksGoals: { response: [] },
      failingCheckIds: ["123", "124"],
      errorPatterns: [
        {
          id: "123",
          description: "Error Pattern #1",
          count: 100,
          firstSeen: new Date("2025-03-24T00:00:00Z"),
        },
        {
          id: "124",
          description: "Error Pattern #2",
          count: 100,
          firstSeen: new Date("2025-03-24T00:00:00Z"),
        },
      ],
    });

    expect(blocks).toMatchSnapshot();
  });

  it("renders failing account state", () => {
    const blocks = createAccountSummaryBlock({
      accountName: "Failing",
      passingChecks: 40,
      passingChecksDelta: 0,
      degradedChecks: 2,
      degradedChecksDelta: 0,
      failingChecks: 8,
      failingChecksDelta: 0,
      hasIssues: true,
      issuesSummary:
        "New degrading or failing checks detected in the last 24h.",
      failingChecksGoals: { response: [] },
      failingCheckIds: ["123", "124", "125"],
      errorPatterns: [
        {
          id: "123",
          description: "Error Pattern #1",
          count: 10,
          firstSeen: new Date("2025-03-24T00:00:00Z"),
        },
      ],
    });

    expect(blocks).toMatchSnapshot();
  });
});
