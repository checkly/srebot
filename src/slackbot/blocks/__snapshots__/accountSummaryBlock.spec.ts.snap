// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountSummaryBlock renders degraded account state 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Account "Degraded" last 24h",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "text": {
        "text": ":white_check_mark: *PASSING*: 45 
 :warning: *DEGRADED*: 5 
:x: *FAILING*: 0 ",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "New degrading or failing checks detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*Impact Analysis:*
No failing checks detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*Top 3 Error Patterns:*
",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "1. \`Error Pattern #1\` (100 times)
      _First seen: <!date^1742774400^{ago}|3/24/2025, 1:00:00 AM>_",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "2. \`Error Pattern #2\` (100 times)
      _First seen: <!date^1742774400^{ago}|3/24/2025, 1:00:00 AM>_",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
    {
      "elements": [
        {
          "text": {
            "emoji": true,
            "text": "Open Checkly Dashboard",
            "type": "plain_text",
          },
          "type": "button",
          "url": "https://app.checklyhq.com/",
        },
        {
          "action_id": "list_failing_checks",
          "text": {
            "emoji": true,
            "text": "List Failing Checks",
            "type": "plain_text",
          },
          "type": "button",
          "value": "123,124",
        },
        {
          "action_id": "list_error_patterns",
          "text": {
            "emoji": true,
            "text": "List Error Patterns",
            "type": "plain_text",
          },
          "type": "button",
          "value": "123,124",
        },
      ],
      "type": "actions",
    },
  ],
}
`;

exports[`accountSummaryBlock renders failing account state 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Account "Failing" last 24h",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "text": {
        "text": ":white_check_mark: *PASSING*: 40 
 :warning: *DEGRADED*: 2 
:x: *FAILING*: 8 ",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "New degrading or failing checks detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*Impact Analysis:*
No failing checks detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*Top 3 Error Patterns:*
",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "1. \`Error Pattern #1\` (10 times)
      _First seen: <!date^1742774400^{ago}|3/24/2025, 1:00:00 AM>_",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
    {
      "elements": [
        {
          "text": {
            "emoji": true,
            "text": "Open Checkly Dashboard",
            "type": "plain_text",
          },
          "type": "button",
          "url": "https://app.checklyhq.com/",
        },
        {
          "action_id": "list_failing_checks",
          "text": {
            "emoji": true,
            "text": "List Failing Checks",
            "type": "plain_text",
          },
          "type": "button",
          "value": "123,124,125",
        },
        {
          "action_id": "list_error_patterns",
          "text": {
            "emoji": true,
            "text": "List Error Patterns",
            "type": "plain_text",
          },
          "type": "button",
          "value": "123",
        },
      ],
      "type": "actions",
    },
  ],
}
`;

exports[`accountSummaryBlock renders healthy account state 1`] = `
{
  "blocks": [
    {
      "text": {
        "emoji": true,
        "text": "Account "Healthy" last 24h",
        "type": "plain_text",
      },
      "type": "header",
    },
    {
      "text": {
        "text": ":white_check_mark: *PASSING*: 50 
 :warning: *DEGRADED*: 0 
:x: *FAILING*: 0 ",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "No issues detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "text": {
        "text": "*Impact Analysis:*
No failing checks detected in the last 24h.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "type": "divider",
    },
    {
      "elements": [
        {
          "text": {
            "emoji": true,
            "text": "Open Checkly Dashboard",
            "type": "plain_text",
          },
          "type": "button",
          "url": "https://app.checklyhq.com/",
        },
      ],
      "type": "actions",
    },
  ],
}
`;
